core/src/autogluon/core/trainer/abstract_trainer.py

```diff
@@ -2335,9 +2390,58 @@ def _train_single_full(
                 total_resources=total_resources,
                 **model_fit_kwargs,
             )
+        if self.callbacks and check_callbacks:
+            self._callbacks_after_fit(model_names=model_names_trained, stack_name=stack_name, level=level)
         self.save()
         return model_names_trained
 
+    def _callbacks_before_fit(
+        self,
+        *,
+        model: AbstractModel,
+        time_limit: float | None,
+        stack_name: str,
+        level: int,
+    ):
+        skip_model = False
+        ts = time.time()
+        for callback in self.callbacks:
+            callback_early_stop, callback_skip_model = callback.before_fit(
+                trainer=self,
+                model=model,
+                logger=logger,
+                time_limit=time_limit,
+                stack_name=stack_name,
+                level=level,
+            )
+            if callback_early_stop:
+                self._callback_early_stop = True
+            if callback_skip_model:
+                skip_model = True
+            if time_limit is not None:
+                te = time.time()
+                time_limit -= te - ts
+                ts = te
+        return skip_model, time_limit
+
+    def _callbacks_after_fit(
+        self,
+        *,
+        model_names: List[str],
+        stack_name: str,
+        level: int,
+    ):
+        for callback in self.callbacks:
+            callback_early_stop = callback.after_fit(
+                self,
+                model_names=model_names,
+                logger=logger,
+                stack_name=stack_name,
+                level=level,
+            )
+            if callback_early_stop:
+                self._callback_early_stop = True
```

Comment:
If there's no side effect to care about with the `.after_fit()`, i.e. you don't care about callbacks state after this call, then you could break early.

```python
    def _callbacks_after_fit(
        self,
        *,
        model_names: List[str],
        stack_name: str,
        level: int,
    ):
        for callback in self.callbacks:
            callback_early_stop = callback.after_fit(
                self,
                model_names=model_names,
                logger=logger,
                stack_name=stack_name,
                level=level,
            )
            if callback_early_stop:
                self._callback_early_stop = True
                break
```

Or if you're feeling functional ;)

```python
    def _callbacks_after_fit(
        self,
        *,
        model_names: List[str],
        stack_name: str,
        level: int,
    ):
        should_stop_itr = (
            callback.after_fit(
                self,
                model_names=model_names,
                logger=logger,
                stack_name=stack_name,
                level=level,
            )
            for callback in self.callbacks
        )
        self._callback_early_stop = any(should_stop_itr)
```

Reply:
My thought process is that some callbacks could simply be logging related, and I wouldn't want a callback that early stops to prevent a logging callback from doing its logging.

The other thing is technically the later callbacks can check if `trainer._callback_early_stop == True` and skip their logic.

This is a good idea though, and I thought about it too while implementing. I've added the following `self.skip_if_trainer_stopped` parameter to callbacks. When true, the callback logic will be skipped when `trainer._callback_early_stop == True`. This should give the best of both worlds.

Reply:
Sounds good! If you want to make it clearer to users, you could make a shallow class `LoggingCallback`, or conversely, `EarlyStoppingCallback` which just internally sets this flag, i.e. user never sets it. I use the pattern a good bit where the flag is actually set as a class variables, e.g. `LoggingCallback` should not be skipped and `EarlyStoppingCallback` can be stopped.

Gives you the freedom to later change this to a dict or things, or whatever, as long as `LoggingCallback` is never skipped and `EarlyStoppimhCallback` is. Also one less parameter for someone to think about

Reply:
That is a great idea, updated the code, I think this is a good practice I should adopt more often to separate user-level args with developer level args (aka those that are only relevant to specify when subclassing)